File: apps/web/src/lib/tag-catalog.ts
import { apiGet } from "@/lib/api";

export type TagCatalogEntry = {
  slug: string;
  label_da: string;
  category: string;
  keywords: string[];
  synonyms: string[];
  auto_suggest: boolean;
  description_da: string | null;
  usage_count: number | null;
};

export type TagSuggestion = {
  slug: string;
  label_da: string;
  confidence: number;
  source: "catalog_keyword" | "catalog_rule" | "llm" | "manual" | string;
  reason_da: string | null;
};

export type SimilarTicket = {
  id: string;
  ticket_number: string;
  title: string;
  status: string;
  score: number;
  match_reasons: string[];
  tags: string[];
};

let catalogCache: TagCatalogEntry[] | null = null;

export async function fetchTagCatalog(options?: {
  includeUsage?: boolean;
}): Promise<TagCatalogEntry[]> {
  const includeUsage = options?.includeUsage ?? true;
  if (catalogCache && includeUsage) {
    return catalogCache;
  }
  const entries = await apiGet<TagCatalogEntry[]>(
    `/api/v1/tags?include_usage=${includeUsage ? "true" : "false"}`,
  );
  if (includeUsage) {
    catalogCache = entries;
  }
  return entries;
}

export async function suggestTagsFromText(text: string): Promise<{
  suggestions: TagSuggestion[];
  suggested_slugs: string[];
}> {
  const params = new URLSearchParams({ text });
  return apiGet(`/api/v1/tags/suggest?${params.toString()}`);
}

export async function fetchSimilarTickets(
  ticketId: string,
  options?: { closedOnly?: boolean; limit?: number },
): Promise<SimilarTicket[]> {
  const params = new URLSearchParams();
  if (options?.closedOnly) {
    params.set("closed_only", "true");
  }
  if (options?.limit) {
    params.set("limit", String(options.limit));
  }
  const qs = params.toString();
  return apiGet<SimilarTicket[]>(
    `/api/v1/tickets/${ticketId}/similar${qs ? `?${qs}` : ""}`,
  );
}

export function formatCatalogOption(entry: TagCatalogEntry): string {
  const usage =
    entry.usage_count != null && entry.usage_count > 0
      ? ` (${entry.usage_count})`
      : "";
  return `${entry.label_da}${usage}`;
}

File: apps/api/src/star_itsm_api/schemas/tag_catalog.py
from pydantic import BaseModel, Field


class TagCatalogEntryRead(BaseModel):
    slug: str
    label_da: str
    category: str
    keywords: list[str] = Field(default_factory=list)
    synonyms: list[str] = Field(default_factory=list)
    auto_suggest: bool = True
    description_da: str | None = None
    usage_count: int | None = None


class TagSuggestionRead(BaseModel):
    """AI-ready tag suggestion — source distinguishes rule vs future LLM."""

    slug: str
    label_da: str
    confidence: float = Field(ge=0.0, le=1.0)
    source: str = Field(
        description="catalog_keyword | catalog_rule | llm | manual",
    )
    reason_da: str | None = None


class TagSuggestResponse(BaseModel):
    suggestions: list[TagSuggestionRead] = Field(default_factory=list)
    suggested_slugs: list[str] = Field(default_factory=list)


class SimilarTicketRead(BaseModel):
    id: str
    ticket_number: str
    title: str
    status: str
    score: float = Field(ge=0.0, le=1.0)
    match_reasons: list[str] = Field(default_factory=list)
    tags: list[str] = Field(default_factory=list)
